db/migrate/20150525081931_remove_duplicate_tokens.rb

```diff
@@ -0,0 +1,28 @@
+class RemoveDuplicateTokens < ActiveRecord::Migration
+  def up
+    # only managed hosts should have tokens
+    Token.where(:id => Token.joins("left outer join hosts on hosts.id = tokens.host_id").
+                             where("hosts.type != 'Host::Managed'").pluck('tokens.id')).delete_all
+
+    # deleting duplicate tokens
+    hosts_with_duplicate_tokens = Token.having('count(*) > 1').group(:host_id).select('host_id')
+    existing_tokens = Host.where(:id => hosts_with_duplicate_tokens).map(&:token).compact
+    if existing_tokens.empty?
+      Token.where(:id => Token.where(:host_id => hosts_with_duplicate_tokens).pluck('tokens.id')).delete_all
+    else
+      Token.where(:id => Token.where('host_id in (?) and id not in (?)', hosts_with_duplicate_tokens, existing_tokens).pluck('tokens.id')).delete_all
```

Comment:
I see a problem hosts_with_duplicate_tokens here, and wonder if the select('host_id') on line 8 needs to be a pluck.  This line generates a query such as:

    SELECT tokens.id FROM "tokens" WHERE (host_id in (NULL,NULL,NULL,NULL,...) and id not in (50007,50012,...)